File: _raw/example1/src/main/java/com/example/grpc/echo/streaming/bidirectional/BidirectionalStreamingServer.java
package com.example.grpc.echo.streaming.bidirectional;

import com.example.grpc.echo.EchoRequest;
import com.example.grpc.echo.EchoResponse;
import com.example.grpc.echo.EchoServiceGrpc;
import com.example.grpc.echo.Server2;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

import java.util.logging.Logger;

public class BidirectionalStreamingServer {

    private static final Logger logger = Logger.getLogger(BidirectionalStreamingServer.class.getName());

    public static void main(String[] args) throws Exception {
        Server2.start(new EchoServiceImpl(), logger);
    }

    static class EchoServiceImpl extends EchoServiceGrpc.EchoServiceImplBase {
        @Override
        public StreamObserver<EchoRequest> bidirectionalStreamingEcho(StreamObserver<EchoResponse> responseObserver) {
            return new StreamObserver<EchoRequest>() {
                @Override
                public void onNext(EchoRequest request) {
                    System.out.println("server next: " + request.getMessage());
                    EchoResponse response = EchoResponse.newBuilder().setMessage("hello " + request.getMessage()).build();
                    responseObserver.onNext(response);
                }

                @Override
                public void onError(Throwable t) {
                    logger.warning("error: " + Status.fromThrowable(t));
                }

                @Override
                public void onCompleted() {
//                    logger.info("Bidirectional stream completed from client side");
                    System.out.println("server completed");
                    responseObserver.onCompleted();
                }
            };
        }
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/methodtypes/streaming/bidirectional/BidirectionalStreamingServer.kt
package com.example.grpc.methodtypes.streaming.bidirectional

import com.example.grpc.EchoRequest
import com.example.grpc.EchoResponse
import com.example.grpc.EchoServiceGrpcKt
import com.example.grpc.echoResponse
import io.grpc.ServerBuilder
import kotlinx.coroutines.flow.Flow
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

object BidirectionalStreamingServer {
  private val logger = Logger.getLogger(BidirectionalStreamingServer::class.java.name)

  @JvmStatic
  fun main(args: Array<String>) {
    val server = ServerBuilder
      .forPort(50051)
      .addService(EchoServiceImpl())
      .build()
      .start()

    logger.info("server started, listening on ${server.port}")

    Runtime.getRuntime().addShutdownHook(
      Thread {
        System.err.println("server is shutting down")
        try {
          server.shutdown().awaitTermination(10, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
          server.shutdownNow()
        }
        System.err.println("server has been shut down")
      }
    )

    server.awaitTermination()
  }

  private class EchoServiceImpl : EchoServiceGrpcKt.EchoServiceCoroutineImplBase() {
    override suspend fun clientStreamingEcho(requests: Flow<EchoRequest>): EchoResponse {
      val responses = mutableListOf<String>()

      requests.collect { request ->
        logger.info("next request: ${request.message}")
        responses.add("hello ${request.message}")
      }

      return echoResponse { message = responses.joinToString(", ") }
    }
  }
}